### PR TITLE
2143: Support selective tags mirroring

### PR DESCRIPTION
**Pull request by erikj79**
We need a way to mirror tags selectively. Here is how I think I want to solve this. We currently have

"branches": Is either a string of one branch pattern or an array of multiple branch patterns
"tags": Takes either "include" or "only", if value is "only", "branches" cannot also be set

Instead of trying to somehow bend the "tags" parameter into taking some kind of patterns, I want to introduce a new parameter "refspecs". The value is an array (or a single string for convenience). Each string constitutes a single refspec that will be given unmodified to a "git push" command (with a + for force if not already present). The configuration becomes a bit more verbose by using this format, but it's also much more powerful. It makes it possible to express more complex mappings using the mirror bot, which may very well be needed when we start using branches more actively.

When setting "refspecs", we do not accept either "branches" or "tags" in the same configuration.

We still pull everything to the local repository in the bot. The refspec is only applied to the push command. Keeping the local repo the same as the source repo reduced the risk of bot configs interfering with each other through conflicting configurations.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2143](https://bugs.openjdk.org/browse/SKARA-2143): Support selective tags mirroring (**Enhancement** - P4)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Zhao Song](https://openjdk.org/census#zsong) (@zhaosongzs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1600/head:pull/1600` \
`$ git checkout pull/1600`

Update a local copy of the PR: \
`$ git checkout pull/1600` \
`$ git pull https://git.openjdk.org/skara.git pull/1600/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1600`

View PR using the GUI difftool: \
`$ git pr show -t 1600`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1600.diff">https://git.openjdk.org/skara/pull/1600.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1600#issuecomment-1894693601)

### CODE REVIEW

File: vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java [hunk 1]
```diff
@@ -649,6 +649,16 @@ public void push(Tag tag, URI uri, boolean force) throws IOException {
         }
     }
 
+    @Override
+    public void push(String refspec, URI uri, boolean force) throws IOException {
+        if (force && !refspec.equals("+")) {
```
Comment (magicus) [inline vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java:654]: Are you sure you did not mean 
```suggestion
        if (force && !refspec.beginsWith("+")) {
```
?

**Review by magicus — APPROVED** [review]
Looks good as far as I can tell.

File: vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java (hunk 1 above)
Comment (edvbld) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java:654]: I would have opted to just skip the `boolean force` parameter. If we are already entrusting the user to supply a correct refspec then they will probably know that they should prefix the refspec with a `+` if they want to force push.

Comment (erikj79) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java]: You are right, I meant beginsWith. But having thought about it, I think Erik has a point. We shouldn't modify the refspec given by the user.

**Review by zhaosongzs — APPROVED** [review]
LGTM

**Comment by erikj79** [comment]
Test comment.

**Review by magicus — APPROVED** [review]

**Pull request closed without merging**